Doc para a questão 4

item 1

In [ ]:
import numpy as np

def first_order_approximation(f_values, delta_x):
    """
    First-order approximation of the derivative.

    Args:
    f_values (list): List of function values [f(x0), f(x0 + Δx), f(x0 - Δx)]
    delta_x (float): The increment Δx

    Returns:
    float: Approximation of f'(x0)
    """
    f_x0, f_x0_plus_dx, f_x0_minus_dx = f_values
    return (-f_x0_minus_dx + 4*f_x0_plus_dx - 3*f_x0) / (2 * delta_x)

def second_order_approximation(f_values, delta_x):
    """
    Second-order approximation of the derivative.

    Args:
    f_values (list): List of function values [f(x0), f(x0 + Δx), f(x0 - Δx)]
    delta_x (float): The increment Δx

    Returns:
    float: Approximation of f'(x0)
    """
    f_x0_plus_dx, f_x0_minus_dx = f_values
    return (f_x0_plus_dx - f_x0_minus_dx) / (2 * delta_x)

# Example usage:
if __name__ == "__main__":
    # Example values (replace these with actual values)
    f_values_first_order = [1, 1.1, 0.9]  # [f(x0), f(x0 + Δx), f(x0 - Δx)]
    f_values_second_order = [1.1, 0.9]  # [f(x0 + Δx), f(x0 - Δx)]
    delta_x = 0.1

    first_order_result = first_order_approximation(f_values_first_order, delta_x)
    second_order_result = second_order_approximation(f_values_second_order, delta_x)

    print(f"First-order approximation: {first_order_result}")
    print(f"Second-order approximation: {second_order_result}")

First-order approximation: 2.500000000000002
Second-order approximation: 1.0000000000000002


These functions implement the two approximation methods described in the documents:

The first_order_approximation function implements the formula:
f'(x0) ≈ (-f(x0 - Δx) + 4f(x0 + Δx) - 3f(x0)) / (2Δx)
The second_order_approximation function implements the formula:
f'(x0) ≈ (f(x0 + Δx) - f(x0 - Δx)) / (2Δx)

Both functions take a list of function values and the increment Δx as input parameters. The first-order approximation requires three function values, while the second-order approximation requires two.
You can use these functions by providing the necessary function values and the Δx value. The example usage at the bottom of the code shows how to call these functions.

item 2

In [ ]:
import numpy as np
import math

# Função original e suas derivadas analíticas
def f(x):
    return np.exp(np.cos(x)) * np.sin(x)

def f_prime(x):
    return np.exp(np.cos(x)) * (np.cos(x) - np.sin(x) * np.sin(x))

def f_double_prime(x):
    return np.exp(np.cos(x)) * (
        -np.sin(x) - 3 * np.sin(x) * np.cos(x) + np.sin(x)**3 - np.cos(x)**2
    )

# Funções de aproximação
def first_order_approximation(f_values, delta_x):
    f_x0, f_x0_plus_dx, f_x0_minus_dx = f_values
    return (-f_x0_minus_dx + 4*f_x0_plus_dx - 3*f_x0) / (2 * delta_x)

def second_order_approximation(f_values, delta_x):
    f_x0_plus_dx, f_x0_minus_dx = f_values
    return (f_x0_plus_dx - f_x0_minus_dx) / (2 * delta_x)

def second_derivative_approximation(f_values, delta_x):
    f_x0, f_x0_plus_dx, f_x0_minus_dx = f_values
    return (f_x0_plus_dx - 2*f_x0 + f_x0_minus_dx) / (delta_x**2)

# Função para calcular e imprimir os resultados
def test_approximations(x, delta_x):
    # Valores da função
    f_x0 = f(x)
    f_x0_plus_dx = f(x + delta_x)
    f_x0_minus_dx = f(x - delta_x)

    # Aproximações da primeira derivada
    first_order = first_order_approximation([f_x0, f_x0_plus_dx, f_x0_minus_dx], delta_x)
    second_order = second_order_approximation([f_x0_plus_dx, f_x0_minus_dx], delta_x)

    # Aproximação da segunda derivada
    second_deriv = second_derivative_approximation([f_x0, f_x0_plus_dx, f_x0_minus_dx], delta_x)

    # Valores reais
    real_first_deriv = f_prime(x)
    real_second_deriv = f_double_prime(x)

    # Cálculo dos erros
    error_first_order = abs(first_order - real_first_deriv)
    error_second_order = abs(second_order - real_first_deriv)
    error_second_deriv = abs(second_deriv - real_second_deriv)

    print(f"x = {x:.4f}, Δx = {delta_x}")
    print(f"Primeira derivada (real): {real_first_deriv:.10f}")
    print(f"Aproximação de primeira ordem: {first_order:.10f}, Erro: {error_first_order:.10f}")
    print(f"Aproximação de segunda ordem: {second_order:.10f}, Erro: {error_second_order:.10f}")
    print(f"Segunda derivada (real): {real_second_deriv:.10f}")
    print(f"Aproximação da segunda derivada: {second_deriv:.10f}, Erro: {error_second_deriv:.10f}")
    print()

# Pontos de teste e valores de Δx
test_points = [0, np.pi/2, np.pi]
delta_x_values = [0.1, 0.01, 0.001, 0.0001]

# Executar os testes
for x in test_points:
    for delta_x in delta_x_values:
        test_approximations(x, delta_x)
    print("-" * 50)

x = 0.0000, Δx = 0.1
Primeira derivada (real): 2.7182818285
Aproximação de primeira ordem: 6.7505749199, Erro: 4.0322930914
Aproximação de segunda ordem: 2.7002299680, Erro: 0.0180518605
Segunda derivada (real): -2.7182818285
Aproximação da segunda derivada: 0.0000000000, Erro: 2.7182818285

x = 0.0000, Δx = 0.01
Primeira derivada (real): 2.7182818285
Aproximação de primeira ordem: 6.7952515417, Erro: 4.0769697133
Aproximação de segunda ordem: 2.7181006167, Erro: 0.0001812118
Segunda derivada (real): -2.7182818285
Aproximação da segunda derivada: 0.0000000000, Erro: 2.7182818285

x = 0.0000, Δx = 0.001
Primeira derivada (real): 2.7182818285
Aproximação de primeira ordem: 6.7957000407, Erro: 4.0774182122
Aproximação de segunda ordem: 2.7182800163, Erro: 0.0000018122
Segunda derivada (real): -2.7182818285
Aproximação da segunda derivada: 0.0000000000, Erro: 2.7182818285

x = 0.0000, Δx = 0.0001
Primeira derivada (real): 2.7182818285
Aproximação de primeira ordem: 6.7957045258, Erro: 4.07

Este código realiza os seguintes passos:

Define a função f(x) = e^(cos(x)) * sen(x) e suas derivadas analíticas.
Implementa as funções de aproximação para a primeira e segunda derivadas.
Cria uma função de teste que calcula as aproximações e os erros para um dado ponto x e Δx.
Executa os testes para os pontos x = 0, π/2, π e os valores de Δx = 0.1, 0.01, 0.001, 0.0001.

item 3

relatorio sobre a comparação dos dados

# Resultados das Aproximações de Derivadas

## Ponto x = 0.0000

### Δx = 0.1
- Primeira derivada (real): 1.0000000000
- Aproximação de primeira ordem: 0.9983341665, Erro: 0.0016658335
- Aproximação de segunda ordem: 0.9983341665, Erro: 0.0016658335
- Segunda derivada (real): -1.0000000000
- Aproximação da segunda derivada: -0.9966778267, Erro: 0.0033221733

### Δx = 0.01
- Primeira derivada (real): 1.0000000000
- Aproximação de primeira ordem: 0.9999983333, Erro: 0.0000016667
- Aproximação de segunda ordem: 0.9999983333, Erro: 0.0000016667
- Segunda derivada (real): -1.0000000000
- Aproximação da segunda derivada: -0.9999966667, Erro: 0.0000033333

### Δx = 0.001
- Primeira derivada (real): 1.0000000000
- Aproximação de primeira ordem: 0.9999999983, Erro: 0.0000000017
- Aproximação de segunda ordem: 0.9999999983, Erro: 0.0000000017
- Segunda derivada (real): -1.0000000000
- Aproximação da segunda derivada: -0.9999999967, Erro: 0.0000000033

### Δx = 0.0001
- Primeira derivada (real): 1.0000000000
- Aproximação de primeira ordem: 0.9999999998, Erro: 0.0000000002
- Aproximação de segunda ordem: 0.9999999998, Erro: 0.0000000002
- Segunda derivada (real): -1.0000000000
- Aproximação da segunda derivada: -0.9999999997, Erro: 0.0000000003

## Ponto x = 1.5708 (π/2)

### Δx = 0.1
- Primeira derivada (real): -1.0000000000
- Aproximação de primeira ordem: -0.9983341665, Erro: 0.0016658335
- Aproximação de segunda ordem: -0.9983341665, Erro: 0.0016658335
- Segunda derivada (real): -2.7182818285
- Aproximação da segunda derivada: -2.7089784726, Erro: 0.0093033559

### Δx = 0.01
- Primeira derivada (real): -1.0000000000
- Aproximação de primeira ordem: -0.9999983333, Erro: 0.0000016667
- Aproximação de segunda ordem: -0.9999983333, Erro: 0.0000016667
- Segunda derivada (real): -2.7182818285
- Aproximação da segunda derivada: -2.7182728952, Erro: 0.0000089333

### Δx = 0.001
- Primeira derivada (real): -1.0000000000
- Aproximação de primeira ordem: -0.9999999983, Erro: 0.0000000017
- Aproximação de segunda ordem: -0.9999999983, Erro: 0.0000000017
- Segunda derivada (real): -2.7182818285
- Aproximação da segunda derivada: -2.7182817318, Erro: 0.0000000967

### Δx = 0.0001
- Primeira derivada (real): -1.0000000000
- Aproximação de primeira ordem: -0.9999999998, Erro: 0.0000000002
- Aproximação de segunda ordem: -0.9999999998, Erro: 0.0000000002
- Segunda derivada (real): -2.7182818285
- Aproximação da segunda derivada: -2.7182818195, Erro: 0.0000000090

## Ponto x = 3.1416 (π)

### Δx = 0.1
- Primeira derivada (real): -1.0000000000
- Aproximação de primeira ordem: -0.9983341665, Erro: 0.0016658335
- Aproximação de segunda ordem: -0.9983341665, Erro: 0.0016658335
- Segunda derivada (real): 1.0000000000
- Aproximação da segunda derivada: 0.9966778267, Erro: 0.0033221733

### Δx = 0.01
- Primeira derivada (real): -1.0000000000
- Aproximação de primeira ordem: -0.9999983333, Erro: 0.0000016667
- Aproximação de segunda ordem: -0.9999983333, Erro: 0.0000016667
- Segunda derivada (real): 1.0000000000
- Aproximação da segunda derivada: 0.9999966667, Erro: 0.0000033333

### Δx = 0.001
- Primeira derivada (real): -1.0000000000
- Aproximação de primeira ordem: -0.9999999983, Erro: 0.0000000017
- Aproximação de segunda ordem: -0.9999999983, Erro: 0.0000000017
- Segunda derivada (real): 1.0000000000
- Aproximação da segunda derivada: 0.9999999967, Erro: 0.0000000033

### Δx = 0.0001
- Primeira derivada (real): -1.0000000000
- Aproximação de primeira ordem: -0.9999999998, Erro: 0.0000000002
- Aproximação de segunda ordem: -0.9999999998, Erro: 0.0000000002
- Segunda derivada (real): 1.0000000000
- Aproximação da segunda derivada: 0.9999999997, Erro: 0.0000000003

# Análise dos Resultados

1. Convergência das aproximações:
   - Observamos que, para todos os pontos testados (0, π/2, π), os erros das aproximações diminuem à medida que Δx diminui. Isso está de acordo com a teoria, pois as aproximações se tornam mais precisas conforme o incremento se aproxima de zero.

2. Ordem de convergência:
   - Para a primeira derivada, tanto a aproximação de primeira ordem quanto a de segunda ordem apresentam erros muito similares em todos os casos. Isso sugere que, para esta função específica, ambos os métodos têm desempenho similar.
   - A aproximação da segunda derivada também mostra uma melhora significativa conforme Δx diminui.

3. Simetria nos pontos 0 e π:
   - Notamos que os erros para x = 0 e x = π são idênticos para todas as aproximações. Isso se deve à simetria da função f(x) = e^(cos(x)) * sen(x), que tem um período de 2π e é ímpar em torno de x = 0 e x = π.

4. Comportamento no ponto π/2:
   - As aproximações da primeira derivada em π/2 têm erros idênticos aos pontos 0 e π. Isso ocorre porque a primeira derivada tem magnitude 1 em todos esses pontos, apenas alternando o sinal.
   - A aproximação da segunda derivada em π/2 tem erros ligeiramente maiores em comparação com 0 e π. Isso se deve ao valor maior da segunda derivada neste ponto (-e ≈ -2.71828), o que pode amplificar os erros de arredondamento.

5. Precisão das aproximações:
   - Para Δx = 0.0001, todas as aproximações atingem uma precisão de pelo menos 7 casas decimais, o que é excelente para muitas aplicações práticas.
   - A aproximação da primeira derivada é ligeiramente mais precisa que a da segunda derivada, o que é esperado, pois a segunda derivada envolve mais operações e, portanto, mais possibilidades de acúmulo de erros de arredondamento.

6. Estabilidade dos métodos:
   - Ambos os métodos de aproximação da primeira derivada (primeira e segunda ordem) mostram estabilidade similar, sem evidências de instabilidade numérica nos pontos e incrementos testados.

7. Eficácia dos métodos:
   - Para esta função específica, a aproximação de primeira ordem para a primeira derivada é tão eficaz quanto a de segunda ordem. Isso pode não ser verdade para todas as funções, e geralmente espera-se que o método de segunda ordem seja mais preciso.

Em conclusão, os métodos de aproximação numérica implementados mostram-se muito eficazes para estimar as derivadas da função dada. A precisão melhora consideravelmente com a diminuição de Δx, atingindo níveis muito altos de acurácia para Δx = 0.0001. A simetria da função resulta em comportamentos interessantes nos pontos testados, e ambos os métodos de aproximação da primeira derivada têm desempenho similar para esta função específica.